File: Q0098_Validate_Binary_Search_Tree/Solution.java
package Q0098_Validate_Binary_Search_Tree;

/**
 * @ Description:
 * @ Date: Created in 22:24 2019-04-14
 * @ Author: Anthony_Duan
 */
class TreeNode {
    int val;
    TreeNode left;
    TreeNode right;

    TreeNode(int x) {
        val = x;
    }
}

public class Solution {
    //定义一个最小的数字
    double last = -Double.MAX_VALUE;
    //标记是否是根节点 根节点要做特殊处理
    boolean isFirst = true;

    public boolean isValidBST(TreeNode root) {

        if (root == null) {
            return true;
        }

        if (isValidBST(root.left)) {
            if (last < root.val || isFirst) {
                isFirst = false;
                last = root.val;
                return isValidBST(root.right);
            }
        }
        return false;
    }

    public static void main(String[] args) {
        TreeNode root = new TreeNode(1);
        root.left = new TreeNode(1);
        System.out.println(new Solution().isValidBST(root));

    }
}

File: Q0098_Validate_Binary_Search_Tree/Solution.py
# Definition for a binary tree node.
class TreeNode:
    def __init__(self, x):
        self.val = x
        self.left = None
        self.right = None


class Solution:
    is_first = True
    last = float('-inf')

    def isValidBST(self, root: TreeNode) -> bool:
        if root is None:
            return True
        if self.isValidBST(root.left):
            if self.last < root.val or self.is_first:
                self.is_first = False
                self.last = root.val
                return self.isValidBST(root.right)
        return False
